apis/python/src/tiledbsoma/_read_iters.py

```diff
@@ -5,59 +5,424 @@
+import pandas as pd
 import pyarrow as pa
 import somacore
+from scipy import sparse
+from somacore import options
+from somacore.query._eager_iter import EagerIterator
 
 # This package's pybind11 code
 import tiledbsoma.pytiledbsoma as clib
 
+from . import _util
+from ._exception import SOMAError
 from ._types import NTuple
 
+if TYPE_CHECKING:
+    from . import SparseNDArray
+
+
+# Convenience types
+_RT = TypeVar("_RT")
+BlockwiseTableReadIterResult = Tuple[pa.Table, Tuple[pa.Array, ...]]
+BlockwiseSingleAxisTableIter = Iterator[BlockwiseTableReadIterResult]
+
+BlockwiseScipyReadIterResult = Tuple[
+    Union[sparse.csr_matrix, sparse.csc_matrix, sparse.coo_matrix],
+    Tuple[npt.NDArray[np.int64], npt.NDArray[np.int64]],
+]
+
+IndicesType = Tuple[npt.NDArray[np.int64], npt.NDArray[np.int64]]
+IJDType = Tuple[
+    Tuple[npt.NDArray[np.int64], npt.NDArray[np.int64]], npt.NDArray[np.generic]
+]
+
 
 class TableReadIter(somacore.ReadIter[pa.Table]):
     """Iterator over `Arrow Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`_ elements"""
 
     def __init__(self, sr: clib.SOMAArray):
-        self.sr = sr
+        self._reader = _arrow_table_reader(sr)
 
     def __next__(self) -> pa.Table:
-        arrow_table = self.sr.read_next()
-        if arrow_table is None:
-            raise StopIteration
-
-        return arrow_table
+        return next(self._reader)
 
     def concat(self) -> pa.Table:
         """Concatenate remainder of iterator, and return as a single `Arrow Table <https://arrow.apache.org/docs/python/generated/pyarrow.Table.html>`_"""
         return pa.concat_tables(self)
 
 
-RT = TypeVar("RT")
+class BlockwiseReadIterBase(somacore.ReadIter[_RT], metaclass=abc.ABCMeta):
+    """Private implementation class. Currently implemented as a single-axis blockwise iterator"""
+
+    _reader: Iterator[_RT]
+
+    def __init__(
+        self,
+        array: "SparseNDArray",
+        sr: clib.SOMAArray,
+        coords: options.SparseNDCoords,
+        axis: Union[int, Sequence[int]],
+        *,
+        size: Optional[Union[int, Sequence[int]]] = None,
+        reindex_disable: Optional[Union[int, Sequence[int], bool]] = None,
+        eager: bool = True,
+        eager_iterator_pool: Optional[ThreadPoolExecutor] = None,
+    ):
+        super().__init__()
+
+        self.ndim = len(sr.shape)
+        self.array = array
+        self.sr = sr
+        self.eager = eager
+        self.eager_iterator_pool = eager_iterator_pool
+
+        # raises on various error checks, AND normalizes args
+        self.axis, self.size, self.reindex_disable = self._validate_args(
+            sr.shape, axis, size, reindex_disable
+        )
+
+        self.major_axis = self.axis[0]
+        self.coords = _pad_with_none(coords, self.ndim)
+
+        # materialize all indexing info.
+        self.joinids: List[pa.Array] = [
+            pa.array(
+                np.concatenate(
+                    list(
```

Comment:
is this `list(` needed? I would expect numpy to take any iterable you give it (though could be wrong)

Reply:
your expectations is wrong - it takes a Sequence, not an interable.

demo:
```
In [4]: np.concatenate( (np.zeros((i,)) for i in range(4)) )
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [4], line 1
----> 1 np.concatenate( (np.zeros((i,)) for i in range(4)) )

File <__array_function__ internals>:200, in concatenate(*args, **kwargs)

TypeError: The first input argument needs to be a sequence

In [5]: np.concatenate( tuple(np.zeros((i,)) for i in range(4)) )
Out[5]: array([0., 0., 0., 0., 0., 0.])
```